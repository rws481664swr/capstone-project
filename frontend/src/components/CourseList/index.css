
.courseListContainer {

}

.CourseList {

}

.courseLi {
    text-decoration: none;
    list-style: none;
}

#CourseList ul {
    padding: 0;
    margin: 12px 10px;

}

#CourseList li {
    /*text-decoration: none;*/
    width:100%;
    padding: 10px;
    border-bottom: 1px solid #ccc;
    margin-bottom: 10px;
    background-color: #f5f5f5;
    border-radius: 5px;
    box-shadow: 0 0 5px #ccc;
    list-style-type: none;
}

#CourseList li:hover {
    background-color: #fff;
    box-shadow: 0 0 5px #aaa;
}

#CourseList li a {

    text-decoration: none;
    color: black;
    display: block;
    width: 100%;
}
#CourseList li a:hover {
    text-decoration: none;
    color: black;

}

#CourseList li{
    display:flex;
    justify-content: space-between;
}

.subject {

    font-size: 20px;
    font-weight: bold;
    color: #000;

}
