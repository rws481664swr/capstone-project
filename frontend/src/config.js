const {  API_URL} = process.env
export let BASE_URL = process.env.DEV_API || 'http://localhost:3001'
if(process.env.NODE_ENV === 'production') BASE_URL = API_URL
