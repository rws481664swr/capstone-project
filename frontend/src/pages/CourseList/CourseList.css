

#join-fab{
    padding: 30px;
    margin-left: 10px;
    display: flex;
    justify-content: space-between;
}

#CourseList ul {
    padding: 0;
    margin: 12px 10px;

}

#CourseList li {
    margin-top: 20px;
    width:100%;
    padding: 10px;
    border-bottom: 1px solid #ccc;
    margin-bottom: 10px;
    background-color: #f5f5f5;
    border-radius: 5px;
    box-shadow: 0 0 5px #ccc;
    list-style-type: none;
    text-decoration: none;
    list-style: none;
}

#CourseList li:hover {
    background-color: #fff;
    box-shadow: 0 0 5px #aaa;
}

#CourseList li a {

    text-decoration: none;
    color: black;
    display: block;
    width: 100%;
}
#CourseList li a:hover {
    text-decoration: none;
    color: black;

}

#CourseList li{
    display:flex;
    justify-content: space-between;
}

.subject {

    font-size: 20px;
    font-weight: bold;
    color: #000;

}
.course-scroll{
    overflow-y: auto;
    height: 100%;
    width: 100%;
}
#join-fab.CourseList_add{
    display: flex;
    justify-content: center;
    align-items: center;
}