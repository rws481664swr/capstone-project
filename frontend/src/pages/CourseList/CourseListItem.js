import {Link} from "react-router-dom";
import useDropCourse from "../../hooks/ajax/useDropCourse";
import {DeleteButton} from "../../components/General/Button/IconButtons/IconButtons";

const CourseListItem = ({drop,course: {courseName, subject, _id, courseNumber, students, ...rest}}) => {

    const dropCourse= drop(useDropCourse(_id))
    return(

        <li className={`CourseListItem`}>
                <Link to={`/courses/${_id}`}>
                    <div className={'CourseListItem_Body'}>
                        <div>
                            <div>{subject}:#{courseNumber} - {courseName} </div>
                            <div>{students.length} Students</div>
                        </div>
                        <div>
                             <DeleteButton onClick={dropCourse}/>
                        </div>
                    </div>
                </Link>
        </li>
)


}
export default CourseListItem

