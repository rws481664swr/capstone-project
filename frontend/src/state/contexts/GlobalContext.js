import {createContext, useContext} from "react";
import jwtDecode from "jwt-decode";
import useLocalStorageState from "../../hooks/useLocalStorageState";

const value = {
    username: undefined,
    role: undefined,
    _id: undefined,
    timestamp: undefined,
    token:null,
    setToken:()=>{}

}
const GlobalContext = createContext(value)
export default GlobalContext

export const GlobalContextProvider = ({children}) => {
    const [context,setContext] =
        useLocalStorageState( 'ctx',value||{})
    const setToken= (token)=>{
        if (token){
            const {

                username,
                role,
                _id,
                timestamp
            } = jwtDecode(token)
            setContext({username, role, _id, timestamp, token})
        }else{
            setContext({})
        }
    }
    const logout= ()=> {
        setToken(null)
    }
    return <GlobalContext.Provider value={{...context,setToken,logout}}>
        {children}
    </GlobalContext.Provider>
}
export const useGlobalContext = () =>     useContext(GlobalContext)

