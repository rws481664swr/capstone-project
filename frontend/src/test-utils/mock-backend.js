import {BASE_URL} from "../config";

export const extractKeysAsList= (obj)=>
    Object.keys(obj).map(key=>obj[key])
        .reduce((acc, val)=>[...acc,val],[])
export const MOCK_TOKEN = {current: 'TOKEN'}

export const spliceUrl = (url) => {
    const arr = BASE_URL.split('')
    const toReturn = url.split('')
    while (arr.length){
        arr.shift()
        toReturn.shift()
    }
    return toReturn.join('')

}