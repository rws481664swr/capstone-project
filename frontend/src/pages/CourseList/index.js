import './CourseList.css'
import CourseListItem from "./CourseListItem";
import {useNavigate} from "react-router-dom";
import useGet from '../../hooks/ajax/useGet'
import FAB from "../../components/General/Button/FAB/FAB";
import {FontAwesomeIcon} from "@fortawesome/react-fontawesome";
import {faPlus} from "@fortawesome/free-solid-svg-icons";
import {useEffect, useState} from "react";
import useFlash from "../../hooks/form/useFlash";

const CourseList = ({scroll = false, displayFAB=true}) => {

    const navigate = useNavigate()
    const [
        toRender, flash, , {danger,success}
    ]= useFlash()
    const gottenCourses = useGet('courses', {init: []})
    const [courses, setCourses] = useState(() => gottenCourses)
    useEffect(() => {
        setCourses(gottenCourses)
    }, [gottenCourses])

    const drop = (_id) => (dropCourse) => async event => {
        event.stopPropagation()
        event.preventDefault()

        try {
            await dropCourse()
            setCourses(courses =>
                courses
                    .filter(course => course._id !== _id))

        } catch (e) {
            console.error(e.message)
        }
    }
    return <>
        <div id={'CourseList'} className={`courseListContainer ${scroll && 'course-scroll'}`}>
            {toRender&& <div>{toRender}</div>}
            <div className="CourseList_Container ">
                <ul  className={'  scrollable'}>
                    {courses && courses.map(course =>
                        <CourseListItem drop={     drop(course._id)} onClick={navigate

                        } key={course._id} course={course}/>)}
                </ul>
            </div>
        </div>
        {displayFAB&&<FAB id={'join-fab'} className={'CourseList_add'} onClick={() => navigate('/join')}>
            <FontAwesomeIcon icon={faPlus}/>
        </FAB>}
    </>
}
export default CourseList

