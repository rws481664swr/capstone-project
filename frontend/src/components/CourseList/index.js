import './index.css'
import CourseListItem from "./CourseListItem";
import {useGlobalContext} from "../../GlobalContext";
import {useNavigate} from "react-router-dom";
import useGet from '../../hooks/useGet'


const CourseList = () => {

    const navigate = useNavigate()
    const courses = useGet('courses',{init:[]})
    if (!courses) return navigate('/login') || <>ERROR</>

    return <>{courses.map((clazz, i) =>
        <CourseListItem key={i} clazz={clazz}/>)}
    </>
}
export default CourseList

