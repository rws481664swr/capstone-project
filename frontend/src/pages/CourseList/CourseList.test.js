import CourseList from "./index";
import {MemoryRouter} from "react-router-dom";
import {act, render} from "@testing-library/react";
import axios from 'axios'
import getData from "../../mock_data/data";
import {default as GlobalContext} from '../../state/contexts/GlobalContext'
import {extractKeysAsList, MOCK_TOKEN} from "../../../mock-backend";
import {BASE_URL} from "../../config";
import chalk from "chalk";
const mockCoursesList = extractKeysAsList(getData().courses)

jest.mock('axios'  ,()=>({
    ...jest.requireActual('axios'),
    get:jest.fn()
}))
const mockGet= (url, cfg) => {

    if(url.includes(`/courses`)){
        return   Promise.resolve({data:mockCoursesList})
    }else if (url.includes(`/auth/token`)){
        return  Promise.resolve({data: {token:MOCK_TOKEN.current}})
    }
    throw new Error( `MOCK GET: ${url} not implemented`)
}
it('should render CourseList', async () => {
    // axios.get.mockImplementation()
    axios.get.mockImplementation(mockGet)


    const fromCourse = ({subject, courseName, courseNumber}) =>
        `${subject}:#${courseNumber} - ${courseName}`
    const {getByText} = await act(() => render(
        <MemoryRouter>
            <GlobalContext.Provider value={{token:MOCK_TOKEN}}>
                <CourseList/>
            </GlobalContext.Provider>
        </MemoryRouter>))
    mockCoursesList.forEach(course => {
        const txt =
            getByText(fromCourse(course))
        expect(txt).toBeInTheDocument()
    })

})
 it('should render CourseList as Fragment', async () => {

     axios.get.mockImplementation(mockGet)



    const {getByText, asFragment} = await act(() => render(
        <MemoryRouter>
            <GlobalContext.Provider value={{token:MOCK_TOKEN}}>
                <CourseList/>
            </GlobalContext.Provider>
        </MemoryRouter>))
    expect(asFragment()).toMatchSnapshot()

})