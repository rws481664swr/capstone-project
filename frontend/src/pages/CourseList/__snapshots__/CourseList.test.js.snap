// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render CourseList as Fragment 1`] = `
<DocumentFragment>
  <div
    class="courseListContainer false"
    id="CourseList"
  >
    <div
      class="CourseList_Container "
    >
      <ul
        class="  scrollable"
      >
        <li
          class="CourseListItem"
        >
          <a
            href="/courses/646e33a759eab5cd39bbb8a9"
          >
            <div
              class="CourseListItem_Body"
            >
              <div>
                <div>
                  subject1:#76351 - c1 
                </div>
                <div>
                  0 Students
                </div>
              </div>
              <div>
                <button
                  class=" DeleteButton IconButton"
                  data-testid="DeleteButton"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-trash "
                    data-icon="trash"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </a>
        </li>
        <li
          class="CourseListItem"
        >
          <a
            href="/courses/646e33a759eab5cd39bbb8aa"
          >
            <div
              class="CourseListItem_Body"
            >
              <div>
                <div>
                  subject2:#24034 - c2 
                </div>
                <div>
                  0 Students
                </div>
              </div>
              <div>
                <button
                  class=" DeleteButton IconButton"
                  data-testid="DeleteButton"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-trash "
                    data-icon="trash"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </a>
        </li>
        <li
          class="CourseListItem"
        >
          <a
            href="/courses/646e33a759eab5cd39bbb8ab"
          >
            <div
              class="CourseListItem_Body"
            >
              <div>
                <div>
                  subject3:#94630 - c3 
                </div>
                <div>
                  0 Students
                </div>
              </div>
              <div>
                <button
                  class=" DeleteButton IconButton"
                  data-testid="DeleteButton"
                >
                  <svg
                    aria-hidden="true"
                    class="svg-inline--fa fa-trash "
                    data-icon="trash"
                    data-prefix="fas"
                    focusable="false"
                    role="img"
                    viewBox="0 0 448 512"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M135.2 17.7L128 32H32C14.3 32 0 46.3 0 64S14.3 96 32 96H416c17.7 0 32-14.3 32-32s-14.3-32-32-32H320l-7.2-14.3C307.4 6.8 296.3 0 284.2 0H163.8c-12.1 0-23.2 6.8-28.6 17.7zM416 128H32L53.2 467c1.6 25.3 22.6 45 47.9 45H346.9c25.3 0 46.3-19.7 47.9-45L416 128z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </a>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="CourseList_add FAB"
    data-testid="fab"
    id="join-fab"
    style="bottom: 1rem; right: 1rem;"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-plus "
      data-icon="plus"
      data-prefix="fas"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 80c0-17.7-14.3-32-32-32s-32 14.3-32 32V224H48c-17.7 0-32 14.3-32 32s14.3 32 32 32H192V432c0 17.7 14.3 32 32 32s32-14.3 32-32V288H400c17.7 0 32-14.3 32-32s-14.3-32-32-32H256V80z"
        fill="currentColor"
      />
    </svg>
  </div>
</DocumentFragment>
`;
